Refactor CreateEffectWarhead ImpactTypes

reaperrr opened the pull request:
Not only were there quite a few of them, more importantly they also ignored target validity.

This PR solves both.
As side-effect, it fixes the issue in RA where a weapon hitting a cell occupied by a submerged sub would display an explosion instead of a water splash.

TD and D2k didn't use custom impact type settings, so all they need are a few yaml changes to cover the fact that `Warhead.ValidTargets` does not contain `Air` by default.

Closes #11848.

reaperrr committed:
Refactor CreateEffectWarhead ImpactTypes

Use target validity to simplify ImpactTypes.

reaperrr committed:
Tweak RA naval vessel target types

The extremely coarse, shared with terrain Ground and Water types make it hard to distinguish between different actor categories, so we add Ship to all non-submerged vessels.
Additionally, we add 'Water' to submerged subs to prevent players from finding them by moving the cursor of a ship with DepthCharge over the suspected sub position.
The latter also allows force-firing DepthCharge at empty water cells.

reaperrr committed:
Adapt RA ballistic weapons to CEWH refactor

reaperrr committed:
Adapt RA smallarms to CEWH refactor

reaperrr committed:
Adapt RA explosions, supers and other weapons to CEWH refactor

reaperrr committed:
Adapt RA missiles to CEWH refactor

reaperrr committed:
Adapt TS effect warheads to CEWH refactor

reaperrr committed:
RA mission weapon warhead fixes

reaperrr committed:
Upgrade rule for CreateEffectWarhead ImpactTypes -> Targets

reaperrr committed:
Adapt TD anti-air weapons to CEWH refactor

reaperrr committed:
Adapt D2k AA missiles to CEWH refactor

GraionDilach commented:
This looks like the AS solution. If a followup then shifts the impacttypes to warheads, then this looks really promising.

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -828,6 +828,53 @@ internal static void UpgradeWeaponRules(ModData modData, int engineVersion, ref
+							Console.WriteLine("CreateEffectWarhead now uses Valid-/InvalidTargets instead of Valid-/InvalidImpactTypes.");
+							Console.WriteLine("Please check whether you need to make manual adjustments.");
+
+							var validTargets = "";
```
It would be cleaner to make this a `List<string>` and then use `FieldSaver.FormatValue` to turn it into a yaml string at the end.

pchote commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -89,6 +69,10 @@ public bool GetDirectHit(World world, CPos cell, WPos pos, Actor firedBy, bool c
 				if (checkTargetType && !IsValidAgainst(victim, firedBy))
 					continue;
 
+				var healthInfo = victim.Info.TraitInfoOrDefault<HealthInfo>();
```
This can simplify down to `if (!victim.Info.HasTraitInfo<HealthInfo>())`

pchote commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -132,10 +116,22 @@ public bool IsValidImpact(WPos pos, Actor firedBy)
 			var impactType = GetImpactType(world, targetTile, pos, firedBy);
-			if (!ValidImpactTypes.HasFlag(impactType) || InvalidImpactTypes.HasFlag(impactType))
-				return false;
+			var validImpact = false;
```
This can be simplify down to:
```csharp
var validImpact = false;
switch (impactType)
{
	case ImpactType.TargetHit:
		return true;
	case ImpactType.Air:
		return IsValidTarget(new string[] { "Air" });
	case ImpactType.Ground:
		var tileInfo = world.Map.GetTerrainInfo(targetTile);
		return IsValidTarget(tileInfo.TargetTypes);
	default:
		return false;
}
```

reaperrr replied:
I assume you mean `switch (validImpact)`?

pchote replied:
Nope, I mean `switch (impactType)` but the `var validImpact` can be deleted.

pchote reviewed (commented):
Here's a couple of code style suggestions to start with.   I haven't tested this ingame, but the yaml changes look like I would expect.

pchote commented:
Needs a rebase now, too.

reaperrr commented:
Rebased and updated.

pchote reviewed (commented):
Weapons hitting boats in RA show both splash and explosion animations.  I don't think this is intentional?

reaperrr commented:
> Weapons hitting boats in RA show both splash and explosion animations. I don't think this is intentional?

Only for DepthCharge vs. submerged subs, not the rest. I'll take a look.

reaperrr commented:
Turns out that the original version of this had a rather significant limitation:

If `GetDirectHit` only found invalid actors, it would return `false`, which would have the same result as if there weren't any actors on that cell at all: `GetImpactType` would return `ImpactType.Ground`. As a result, even if there were 'blacklisted', invalid actors on that cell, as long as the cell itself contained a valid target type, the warhead would still trigger.
Playing the animation + sound even when a cell contains explicitly invalid actors is counter-intuitive, and broke RA, since `Water` is obviously valid for splashes, so they would play despite `Ship` being blacklisted.

I fixed this by upgrading the `GetDirectHit` boolean to `GetDirectHitTargetType` using a new `ImpactTargetType` enum with the types `ValidActor`, `InvalidActor` and `NoActor`, and returning `ImpactType.None` in `GetImpactType` if `ImpactTargetType` is `InvalidActor`.

Additionally, I forgot to black-list `Structure` for splashes in RA, so it would have played on ship yard/sub pen even when the code was fixed.

I'm pretty confident this now works as intended and covers all edge cases (at least it does in RA, which is our most complex test-case for this), but you never know, so review carefully.

RoosterDragon commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -56,49 +53,53 @@ void IRulesetLoaded<WeaponInfo>.RulesetLoaded(Ruleset rules, WeaponInfo info)
+				if (!directHit)
+					continue;
+
+				if (directHit && (!checkTargetValidity || (checkTargetValidity && IsValidAgainst(victim, firedBy))))
```
Simplifies to `if (!checkTargetValidity || IsValidAgainst(victim, firedBy))`.

RoosterDragon commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -143,10 +144,18 @@ public bool IsValidImpact(WPos pos, Actor firedBy)
+				case ImpactType.TargetHit:
+					return true;
+				case ImpactType.Air:
+					return IsValidTarget(new string[] { "Air" });
```
Consider caching this array.

pchote commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -54,51 +51,57 @@ void IRulesetLoaded<WeaponInfo>.RulesetLoaded(Ruleset rules, WeaponInfo info)
 				VictimScanRadius = Util.MinimumRequiredVictimScanRadius(rules);
 		}
 
+		readonly string[] targetTypeAir = new string[] { "Air" };
```
Style nit: `static readonly string[] TargetTypeAir = { "Air" };`

pchote commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -54,51 +51,57 @@ void IRulesetLoaded<WeaponInfo>.RulesetLoaded(Ruleset rules, WeaponInfo info)
+				var targetType = GetDirectHitTargetType(world, cell, pos, firedBy, true);
+				if (targetType == ImpactTargetType.ValidActor)
+					return ImpactType.TargetHit;
+				else if (targetType == ImpactTargetType.InvalidActor)
```
Style nit: you can omit the `else` here (use `if` by itself) because the first branch returns.

pchote commented on OpenRA.Mods.Common/Warheads/CreateEffectWarhead.cs:
```diff
@@ -54,51 +51,57 @@ void IRulesetLoaded<WeaponInfo>.RulesetLoaded(Ruleset rules, WeaponInfo info)
 
-			return false;
+			// There was at least a single direct hit, but none on valid target(s)
+			if (invalidHit)
```
Style nit: `return invalidHit ? ImpactTargetType.InvalidActor : ImpactTargetType.NoActor;`

pchote reviewed (commented):
A couple of minor style nits for you while I test this ingame.

pchote commented:
* Planes/helicopters crashing into the ground/water disappear instead of exploding.
* Killed boats show both an explosion and splash.

reaperrr commented:
Updated.

The boats showing both as well as the missing flak cannon air explosion (noticed by myself) were just yaml issues.

> Planes/helicopters crashing into the ground/water disappear instead of exploding.

This is basically a pecularity of aircraft husks: since they are untargetable, but *do* have `Health` and `HitShape` (which are both needed to make `self.Kill` work in `FallToEarth`), the victim check would never `continue` early (impact pos will obviously always be within their hitshape) and would always return `InvalidActor` on them.
Added
```
				if (!AffectsParent && victim == firedBy)
					continue;
```
to the beginning of the victim check loop, which also has the nice default side-effect of probably improving performance a little if the `firedBy` actor is within the `VictimScanRadius`.

jrb0001 commented:
The hardcoded value "Air" should be mentioned in the documentation for `ValidTargets` and `InvalidTargets`. An even better way would be to make it configurable, but I don't see how to do that properly at right now.

As Graion has already mentioned, this logic should be moved to `Warhead` to reuse it in other warheads. A followup for that might be a good idea to avoid scope creep of this PR.

reaperrr commented:
> The hardcoded value "Air" should be mentioned in the documentation for ValidTargets and InvalidTargets.

Doing that now would misleading, because it's only true for the CEWH (for now). I'll add a comment about it to `AirThreshold`'s desc, though.

reaperrr committed:
Mention hard-coded 'Air' TargetType in CEWH.AirThreshod desc

GraionDilach commented:
I had thought about this during the time I set up the WarheadAS class back then, but my other idea (which I felt messy in a different way) was to exposing the targettypes of air through expanding all theater yamls with an Air terraintype - which would shift the hardcode to a separate level.

reaperrr commented:
Updated.

> I had thought about this during the time I set up the WarheadAS class back then, but my other idea (which I felt messy in a different way) was to exposing the targettypes of air through expanding all theater yamls with an Air terraintype - which would shift the hardcode to a separate level.

I think compared to that, simply adding an `AirTargetType(s)` field to Warhead would be the lesser evil.

jrb0001 commented:
The description looks good to me.

> I think compared to that, simply adding an `AirTargetType(s)` field to Warhead would be the lesser evil.

Wouldn't it be better to add a boolean like `HitEmptyAir` instead? Adding a new field just to compare it with another field sounds a bit weird to me.

abcdefg30 commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -900,6 +900,53 @@ internal static void UpgradeWeaponRules(ModData modData, int engineVersion, ref
+						else
+						{
+							// 'Air' is not part of the internal warhead ValidTargets default, but was part of the ValidImpactTypes default.
+							node.Value.Nodes.Add(new MiniYamlNode("ValidTargets", "Ground, Water, Air"));
```
What if a `ValidTargets` yaml node already exists?

reaperrr replied:
Normally it shouldn't, since ValidTargets previously had no effect for CEWH, but I'll add a safety-net just in case.

abcdefg30 commented:
Depth charges don't damage surfaced subs, I think that should be changed.

reaperrr commented:
> Depth charges don't damage surfaced subs, I think that should be changed.

It's like that on bleed too, though. Are you sure?

abcdefg30 commented:
Yes, as you can dodge depth charges by force firing with the submarine, which looks bogus to me. But if it's already that way on bleed, you can leave it as is, I just thought you touch that anyway in this PR.

pchote commented:
If that is a regression vs `release-20170527` then we will still need to fix it before the playtest.

reaperrr commented:
> If that is a regression vs release-20170527 then we will still need to fix it before the playtest.

It isn't, it's the same in last release. 
I agree with @abcdefg30 that it looks bogus, but it would definitely affect balance to fix it now, since DepthCharge is stronger vs. subs than the Gunboat cannon.

reaperrr commented:
Updated upgrade rule.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/aad16a0588898957ab59b987c5b2946e7b59e587...fbeba2c5a7c09b72d0a773abd67484ee50b16068)
